text/0003-templates.md

```diff
@@ -0,0 +1,258 @@
+
+While there are situations differing versions are intentional, this is more often accidental. Multiple differing versions arise from not reviewing `pnpm-lock.yaml` file changes or not searching for existing dependency specifiers before adding a new one. The later is typically unwritten convention in most monorepos.
+
+### Authoring Fields
+
+Fields such as `author`, `license`, and `repository` are typically expected to be the same across all packages in a monorepo. It would be easier to set this in a singular source of truth rather than every package.
+
+## Detailed Explanation
+
+A "*Template*" is defined as a normal `package.json` file with the `pnpm.template` field set.
+
+```json5
+{
+  "name": "@example/react-template",
+  "pnpm": {
+    "template": true,
+  },
+  "version": "0.1.0",
+  "author": "Example Team <team@team.example>",
+  "license": "MIT",
+  "scripts": {
+    "compile": "tsc"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+Templates must opt into being a template by defining `pnpm.template`. This is a safety mechanism to:
+
+1. Intentionally prevent existing packages from being referred to for that purpose. Templates have different obligations and usage than standard packages that authors of templates should be aware of.
+2. Allow checks to be performed on the template before publishing and during consumption. Not all `package.json` fields may be valid on a template.
+
+A package can reference the template to populate different portions of its definition. The specific feature is determined by `pnpm.templates`: `extends`, `toolkit`, and `catalog`.
+
+### Extends
+
+Using the `pnpm.templates.extends` mechanism, `package.json` fields from templates are copied with small exceptions. The `name`, `dist`, and underscore prefixed fields (e.g. `_npmUser`) are not copied since they're typically specific to the template package itself.
+
+A package referencing the `@example/react-template` above will have the following on-disk and in-memory representations.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "extends": ["@example/react-template@0.1.0"],
+    }
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "version": "0.1.0",
+  "author": "Example Team <team@team.example>",
+  "license": "MIT",
+  "scripts": {
+    "compile": "tsc"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+Note that non-standard `package.json` fields such as `eslintConfig` will also be copied. Although these fields have no meaning to package managers, they will appear in the published manifest.
+
+### Toolkit
```

Comment:
This is meant for the proxy packages mentioned in https://github.com/pnpm/rfcs/pull/1#issuecomment-1399394933.

Reply:
do we need toolkits if we can use extend to copy the dev dependencies?

Reply:
Perhaps. The case I was worried about was toolkits coming from different repositories and defining `author`, `repository`, and `description` fields that the consuming package wouldn't want to inherit through `extends`.

The toolkit template could get around that by not defining any fields other than its `dependencies`, but I think that would be a strange and surprising restriction.

Your question is a good one to be asking. It does feels like there's some overlap between `extends` and `toolkit` that could be reduced.

Reply:
One way we could eliminate `toolkit` is by:

- Adding the concept of `private`/`public` modifiers so fields like `author` can be used safely. https://github.com/pnpm/rfcs/pull/3#discussion_r1127456524
- Requiring the dependencies the toolkit brings in to be `devDependencies` in the first place, so there isn't a strange `dependencies` to `devDependencies` rewrite.

Do you know if the second requirement would be problematic for the users asking for these toolkit-like setups?

Reply:
It sounds like for now we'll keep the `toolkit` flavor and avoid an `extends` template.

- Change initially pushed for in https://github.com/pnpm/rfcs/pull/3#issuecomment-1493167623
- Change pushed in commit c1f3f2aa6c8120c495e4a3ced884dcb445e54085.

Going to hit the "_Resolve conversation_" button for this thread.